ZENKO-933 bugfix: redis chart

ssalaues opened the pull request:
Review only for now since we should thoroughly test this before merging.

Refactor of the redis-ha chart. This fixes issues that would cause redis not to fail over at all or too late.

Commit:
Import stable/redis-ha chart v2.2.1

commit aedcdbb37133e3f5f97fcc997693f46449b31996

Commit:
bugfix: productize redis chart

The upstream redis-ha chart is not reliable nor fault tolerant
causing system failures in an otherwise highly available environment

Commit:
chore: update redis links

bert-e commented:
# Hello ssalaues,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get
information on this process.





*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `bugfix/ZENKO-933-redis-chart` and the
destination branch `development/0.9`.

Please resolve the conflict on **the feature branch** (`bugfix/ZENKO-933-redis-chart`).

```bash
 $ git fetch
 $ git checkout origin/bugfix/ZENKO-933-redis-chart
 $ git merge origin/development/0.9
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:bugfix/ZENKO-933-redis-chart
```

ssalaues commented:
@bert-e wait

rahulreddy commented on charts/redis-ha/templates/redis-ha-configmap.yaml:
```diff
@@ -0,0 +1,29 @@
+
+  sentinel.conf: |
+    dir /data
+    sentinel monitor {{ .Values.redis.masterGroupName }} {{ template "redis-ha.fullname" . }}-0.{{ template "redis-ha.fullname" . }} 6379 {{ .Values.sentinel.quorum }}
```
I can't tell if we are getting the bootstraplist here or just the first host.

ssalaues replied:
I copied Federation which initializes the conf with the first host as the master and then configures the rest as slaved by default.

rahulreddy replied:
Yeah actually that's true. We are good then.

rahulreddy commented on charts/redis-ha/templates/redis-ha-statefulset.yaml:
```diff
@@ -0,0 +1,223 @@
+                    sed -i "s/*.sentinel known.*$OLD_IP//" $SENTINEL_CONF
+                fi
+                if [[ `grep "sentinel monitor" $SENTINEL_CONF` ]]; then
+                    sed -i "s/^.*sentinel monitor.*/sentinel monitor {{ .Values.redis.masterGroupName }} $1 6379 {{ .Values.sentinel.quorum }}/" $SENTINEL_CONF
```
Use the var `.Values.redis.port` instead of hardcoding the port.

rahulreddy commented on charts/redis-ha/templates/redis-ha-statefulset.yaml:
```diff
@@ -0,0 +1,223 @@
+
+            set -ex
+            function sentinel_update(){
+                if [[ "$OLD_IP" != "$POD_IP" ]]; then
```
Why do you have to sed the old ip with the new ones? It should be able to pickup new valid ips and just blacklist old invalid ips?

ssalaues replied:
You're right they should be able to blacklist. I'll remove it and test.

rahulreddy commented on charts/redis-ha/values.yaml:
```diff
@@ -0,0 +1,102 @@
+## Redis specific configuration options
+redis:
+  port: 6379
+  masterGroupName: mymaster
```
I think it's ok if we switch master name to `zenko`. Sentinel should allow other slaves to update their configuration and interpret this as the new master name. This should be tested obviously.

ssalaues replied:
I tested this and didn't work (seamlessly at least) because kubernetes updates the pods in reverse order. So when the redis-ha-2 goes down and comes back up, it tries to contact the master in the `zenko` group that does not exist and goes into crashloop state preventing the remaining pods from updating. However deleting all the pods manually does result in the intended group change.

The default that we would be upgrading from on previous chart is `mymaster` so it might be safest to leave it like this.

rahulreddy replied:
Ah damn, thanks for testing. Let's leave it as-is.

rahulreddy reviewed (commented):
LGTM in general. I added a few comments where I had concerns. There are a few references where port numbers like 6379 are hardcoded, can they be amended to use values?

giacope reviewed (approved):
/lgtm

Commit:
rf: Improve init script

Found many corner cases caused only after several helm upgrades where old masters
come back up and not update their configs causing partitions between the old and
the new masters. Unfortunately partially due to the fact that pods will not retain their
IPs after updates/deletes/restarts. This should be handled correctly now in the init script.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @ssalaues

* @vrancurel

* @giacomoguiulfo

* @NicolasT

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/0.9`




There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of Release Engineering will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of Release Engineering now.

dora-korpar commented on charts/redis-ha/README.md:
````diff
@@ -0,0 +1,82 @@
+$ helm install stable/redis-ha
+```
+
+By default this chart install one master pod containing redis master container and sentinel container, 2 sentinels and 1 redis slave.
````
installs

dora-korpar commented on charts/redis-ha/README.md:
````diff
@@ -0,0 +1,82 @@
+$ helm install stable/redis-ha
+```
+
+The command deploys Redis on the Kubernetes cluster in the default configuration. By default this chart install one master pod containing redis master container and sentinel container along with 2 redis slave pods each containing their own sentinel sidecars. The [configuration](#configuration) section lists the parameters that can be configured during installation.
````
installs 😬

ssalaues replied:
Welp looks like @wabernatScality is gonna have his work cut out for him on this one

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/0.9`




Please check the status of the associated issue ZENKO-933.

Goodbye ssalaues.
